File: army/src/main/java/com/rolan/army/SingleRequest.java
package com.rolan.army;

import android.app.Activity;
import android.content.Context;

import java.util.List;

/**
 * Created by Rolan.Wang on 1/29/21.2:39 PM
 * 描述：
 */
public class SingleRequest implements Request, ResourceCallback {
    private Engine engine;
    private ArmyContext armyContext;
    private Context context;
    private List<RequestListener> requestListeners;
    private BaseRequestOptions<?> requestOptions;

    public SingleRequest(Context context, Engine engine, ArmyContext armyContext, List<RequestListener> requestListeners, BaseRequestOptions<?> requestOptions) {
        this.context = context;
        this.engine = engine;
        this.armyContext = armyContext;
        this.requestListeners = requestListeners;
        this.requestOptions=requestOptions;
    }

    public static SingleRequest obtain(Context context,
                                       Engine engine,
                                       ArmyContext armyContext,
                                       List<RequestListener> requestListeners,
                                       BaseRequestOptions<?> requestOptions) {
        return new SingleRequest(context, engine, armyContext, requestListeners,requestOptions);
    }

    @Override
    public boolean isComplete() {
        return false;
    }

    @Override
    public void recycle() {

    }

    @Override
    public void begin() {
        engine.post(requestOptions.getTaskName(),requestOptions.getPriority(),this);
    }

    @Override
    public void clear() {
        engine.release(requestOptions.getTaskName());
    }

    @Override
    public void canShow() {
        if (requestListeners != null&&context!=null) {
            if (context instanceof Activity){
                if(((Activity) context).isFinishing()){
                    ArmyLog.d("页面已经销毁，不再回调");
                    return;
                }
            }
            for (RequestListener listener : requestListeners) {
                listener.canShow();
            }
        }
    }
}

File: army/src/main/java/com/rolan/army/SingleRequest.kt
package com.rolan.army

import android.app.Activity
import android.content.Context
import com.rolan.army.ArmyLog.d

/**
 * Created by Rolan.Wang on 1/29/21.2:39 PM
 */
class SingleRequest(private val context: Context?, private val engine: IEngine, private val armyContext: ArmyContext?, private val requestListeners: List<RequestListener>?, private val requestOptions: BaseRequestOptions<*>) : Request, ResourceCallback {
    override val isComplete: Boolean
        get() = false

    override fun recycle() {}
    override fun begin() {
        engine.post(requestOptions.taskName, requestOptions.priority, this)
    }

    override fun clear() {
        engine.release(requestOptions.taskName)
    }

    override fun canShow(newTask: SingleTask?) {
        if (requestListeners != null && context != null) {
            if (context is Activity) {
                if (context.isFinishing) {
                    d("page has destory")
                    return
                }
            }
            for (listener in requestListeners) {
                listener.startTask(newTask)
            }
        }
    }

    companion object {
        fun obtain(context: Context?,
                   engine: IEngine,
                   armyContext: ArmyContext?,
                   requestListeners: List<RequestListener>?,
                   requestOptions: BaseRequestOptions<*>): SingleRequest {
            return SingleRequest(context, engine, armyContext, requestListeners, requestOptions)
        }
    }
}